docs: Explanation of shard-awareness (#19)
scylladb/cpp-driver: topics/README.md

  /* Add contact points */
  cass_cluster_set_contact_points(cluster, "127.0.0.1");

  /* Shard-awareness (Scylla only): choose the local (ephemeral) port range */
  cass_cluster_set_local_port_range(cluster, 49152, 65535);

>   /* It is beneficial to set this number to a multiply of node's shard count */

Reviewer: Also, why is it beneficial? Maybe it's worth explaining it here (briefly).

>   /* It is beneficial to set this number to a multiply of node's shard count */

Author: Done + I chose not to dive into it and just described what will happen.

## Cluster

The [`CassCluster`] object describes a Cassandra/Scylla cluster’s configuration.
The default cluster object is good for most clusters and only requires a single
or multiple list of contact points in order to establish a session connection.
Once a session is connected using a cluster object its configuration is
constant. Modifying the cluster object configuration once a session is
established does not alter the session's configuration.

## Session

The [`CassSession`] object is used for query execution. Internally, a session
object also manages a pool of client connections to Cassandra/Scylla and uses
a load balancing policy to distribute requests across those connections. An

> application should create a single session object per keyspace as a session

Reviewer 2: ```suggestion
application should create a single session object per keyspace. A session
```

object is designed to be created once, reused, and shared by multiple threads
within the application. The throughput of a session can be scaled by
increasing the number of I/O threads. An I/O thread is used to handle reading
and writing query request data to and from Cassandra/Scylla. The number of I/O
threads defaults to one per CPU core, but it can be configured using
[`cass_cluster_set_num_threads_io()`]. It’s generally better to create a single
session with more I/O threads than multiple sessions with a smaller number of

> I/O threads. More DataStax driver best practices can be found in this [post].

Reviewer 2: ```suggestion
I/O threads. 
```

## Shard-awareness

> Scylla is built around the concept of "sharded architecture". What it means for

Reviewer 2: ```suggestion
Scylla is built around the concept o a *sharded architecture*. What it means for
```

## Cluster

The [`CassCluster`] object describes a Cassandra/Scylla cluster’s configuration.
The default cluster object is good for most clusters and only requires a single
or multiple list of contact points in order to establish a session connection.
Once a session is connected using a cluster object its configuration is
constant. Modifying the cluster object configuration once a session is

> established does not alter the session's configuration.

Author: That's how they wrote it in the original docs.
Anyway, I'm afraid I don't understand. Do you mean that changes to the Cluster object are taken into account when the next session is opened?

## Session

The [`CassSession`] object is used for query execution. Internally, a session
object also manages a pool of client connections to Cassandra/Scylla and uses
a load balancing policy to distribute requests across those connections. An
application should create a single session object per keyspace as a session
object is designed to be created once, reused, and shared by multiple threads
within the application. The throughput of a session can be scaled by
increasing the number of I/O threads. An I/O thread is used to handle reading
and writing query request data to and from Cassandra/Scylla. The number of I/O
threads defaults to one per CPU core, but it can be configured using
[`cass_cluster_set_num_threads_io()`]. It’s generally better to create a single
session with more I/O threads than multiple sessions with a smaller number of
I/O threads. More DataStax driver best practices can be found in this [post].

## Shard-awareness

> Scylla is built around the concept of "sharded architecture". What it means for

Reviewer 3: How about having an example snippt of code ? Of how shard aware should be used ?

Author: I think it's there, in the modification of the example code. Maybe not clear enough (?)
BTW. thanks a lot for chiming in (!)

Reviewer 3: BTW, I guess you seen that part in the python driver docs: https://python-driver.docs.scylladb.com/stable/scylla_specific.html#shard-awareness

maybe it's not the best example, but something in that direction. 

I don't think we a very long explanation, and we should stick most of it to scylla-core, since it's shard among the drivers, we don't need to keep explaining shard aware again and again in each driver repo

also link to Alexys videos on the the python-driver, might be also helpful
https://github.com/scylladb/python-driver/issues/92 (open an issue to myself to do exactly that)

Author: I've been thinking about it but, sadly, I don't know of a one-to-rule-them-all resource at scylla core. Maybe we should write an ultimate blogpost about shard-awareness?

### "Advanced" shard-awareness

Since Scylla 4.3 however, drivers can use a new, more powerful method of
establishing per-shard connection pools. This is the recommended usage pattern,
commonly referred to as "advanced" shard-awareness. The idea behind it is that
Scylla listens for CQL connections on additional port, by default 19042.
Connections incoming to that port, are being routed to the shard determined by
*client-side (ephemeral) port number*. Precisely, if a client socket has local
port number `P` then that connection lands on shard `P % shard_count`. The
function of the usual port 9042 (`native_transport_port`) is unchanged and
non-shard-aware drivers should continue using it.

**NOTE:** It's important to unblock `native_shard_aware_transport_port` and/or
`native_shard_aware_transport_port_ssl` in the firewall rules.

**NOTE:** If the client app runs behind a NAT (e.g. on a desktop in the office
network) while the Scylla cluster is hosted somewhere else (e.g. on Azure or
AWS) then, most likely, the router at the office alters the client-side port
numbers. In this case port-based ("advanced") shard selection will not work and

> will fall back to the "basic" mode.

Reviewer 3: Is this fallback automatic ?

Author: yes

> will fall back to the "basic" mode.

Reviewer: Sections about "basic" and "advanced" shard awareness are missing an explanation _why_ the advanced method is better than the basic one.

Author: Done

[cpp-driver-dependencies-centos7]: http://downloads.datastax.com/cpp-driver/centos/7/dependencies/
[cpp-driver-dependencies-centos8]: http://downloads.datastax.com/cpp-driver/centos/8/dependencies/
[cpp-driver-dependencies-ubuntu14-04]: http://downloads.datastax.com/cpp-driver/ubuntu/14.04/dependencies/
[cpp-driver-dependencies-ubuntu16-04]: http://downloads.datastax.com/cpp-driver/ubuntu/16.04/dependencies/
[cpp-driver-dependencies-ubuntu18-04]: http://downloads.datastax.com/cpp-driver/ubuntu/18.04/dependencies/
[cpp-driver-dependencies-windows]: http://downloads.datastax.com/cpp-driver/windows/dependencies/
[built from source]: http://github.com/scylladb/cpp-driver/tree/master/topics/building/
[prepared statements]: http://github.com/scylladb/cpp-driver/tree/master/topics/basics/prepared_statements/
[`cass_int32_t`]: http://github.com/scylladb/cpp-driver/tree/master/api/cassandra.h/#cass-int32-t
[`cass_result_first_row()`]: http://github.com/scylladb/cpp-driver/tree/master/api/struct.CassResult/#cass-result-first-row
[`cass_cluster_set_num_threads_io()`]: http://github.com/scylladb/cpp-driver/tree/master/api/struct.CassCluster/#function-cass_cluster_set_num_threads_io

> [`CassCluster`]: http://github.com/scylladb/cpp-driver/tree/master/api/struct.CassCluster/

Reviewer 3: all those links are broken, it's links to autogenerated docs

check with @dgarcia360 how where to point those links, and how

Author: Ooops. For now I'm reverting to datastax docs.

Reviewer 2: @jul-stas - let's get this merged so I have a stable version and then I can make other changes to the links
